src/main/java/com/emailcrawler/crawlerutil/DomainCrawler.java

```diff
@@ -0,0 +1,163 @@
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+
+import javax.mail.internet.AddressException;
+import javax.mail.internet.InternetAddress;
+
+/**
+ * Helper to scan all email address found on a website. Crawls all pages of the
+ * specific domain looking for email addresses. Uses the jsoup parser to parse
+ * pages for links and emails.
+ */
+public class DomainCrawler {
+	// Do not god beyond MAX_DEPTH when following links
+	private static final int MAX_DEPTH = 25;
+	// Base URL to start from
+	private String rootUrl;
+	// Domain name the base url is mapped to
+	private String domainName;
+	// Placeholder for links scanned
+	private HashSet<String> links;
+	// Placeholder for emails found
+	private HashMap<String,String> emails;
+	// The selector to match links to the specific domain
+	private String linkSelector;
+
+	/**
+	 * DomainCrawler Crawl a domain recursively for email addresses.
+	 * 
+	 * @param url
+	 */
+	public DomainCrawler(String url) {
+		rootUrl = getAbsoluteUrl(url);
+		init();
+	}
+
+	private void init() {
+
+		links = new HashSet<String>();
+		emails = new HashMap<String,String>();
+		this.domainName = getDomainName(rootUrl);
+		linkSelector = "a[href*=" + domainName + "]";
+	}
+
+	/**
+	 * Fetch all the emails of the domain
+	 * 
+	 * @return
+	 */
+	public List<String> getEmails() {
+
+		init();
+
+		if (StringUtil.isBlank(domainName)) {
+			System.err.println("Invalid domain name for url " + rootUrl);
+		} else {
+			getEmailsFromPage(rootUrl, 0);
+		}
+		System.out.println("===========>");
+		List<String> emailList = new ArrayList<String>(emails.keySet());
+		for (String email : emailList) {
+			System.out.println(email);
+		}
+		System.out.println("===========> Scanned " + links.size() + " pages.  Found " + emailList.size() + " unique emails ");
+
+		return emailList;
+	}
+
+	/**
+	 * Recursively crawl for pages until a specified depth is reached.
+	 * 
+	 * @param currentURL
+	 * @param depth
+	 */
+	private void getEmailsFromPage(String currentURL, int depth) {
+
+		String domain = getDomainName(currentURL);
+
+		if ((domainName.equals(domain)) && (!links.contains(currentURL)) && (depth < MAX_DEPTH)) {
+			System.out.println(">> Depth: " + depth + " [" + currentURL + "]");
+			try {
+				depth++;
+
+				links.add(currentURL);
+
+				Document document = Jsoup.connect(currentURL).get();
+
+				Elements emailListOnPage = document.select("a[href*=mailto]");
+				Elements linkListOnPage = document.select(linkSelector);
+
+				for (Element page : linkListOnPage) {
+					String newUrl = getAbsoluteUrl(page.attr("abs:href"));
+					getEmailsFromPage(newUrl, depth);
+				}
+
+				for (Element page : emailListOnPage) {
+					//System.out.println("Email " + page.attr("abs:mailto") + " , text " + page.text());
+					String email = page.text();
+					if (isValidEmailAddress(email)) {
+						emails.put(email,email);
+					}
+				}
+
+			} catch (IOException e) {
```

Comment:
What can lead to these exceptions? Safe to proceed still?

Reply:
They could be invalid links which could be valid to ignore (404).
Some of them could be accessible only to a authenticated/authorized user (401/403), in which case we could consider code to login to the domain and then crawl.
There could be timeout errors from a slower response from  the server. This could be either a server that is slow or is using a rate limiter to not allow too many connections. For the later, we could use a strategy to slow down the client.

The current code is throwing an exception if the content type is anything other than text / xml. We could use ignoreContetntType() to download all content, like pdf and write custom code to parse such types.

I think we should consider as many cases as possible and ignore the rest. However, these exceptions should be logged and monitored and custom code should be added  to cover exception cases, as we see them.